File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/jsx-handler-names.ts
import { compileGlob } from "../../utils/compile-glob.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getJsxAttributeName } from "../../utils/get-jsx-attribute-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { Rule } from "../../utils/rule.js";

const buildHandlerNameMessage = (handlerName: string, propKey: string): string =>
  `The handler "${handlerName}" does not match the "${propKey}" event prop convention, so readers cannot trace event flow quickly.`;
const buildHandlerPropMessage = (propKey: string, propValue: string): string =>
  `The prop "${propKey}" passes handler "${propValue}" but is not named like an event prop, so callers cannot tell it fires an event.`;

interface JsxHandlerNamesSettings {
  checkInlineFunction?: boolean;
  checkLocalVariables?: boolean;
  eventHandlerPrefix?: string | false;
  eventHandlerPropPrefix?: string | false;
  ignoreComponentNames?: ReadonlyArray<string>;
}

const DEFAULT_HANDLER_PREFIX = "handle";
const DEFAULT_HANDLER_PROP_PREFIX = "on";

interface ResolvedSettings {
  checkInlineFunction: boolean;
  checkLocalVariables: boolean;
  eventHandlerPrefix: string;
  eventHandlerPropPrefix: string;
  ignoreComponentNames: ReadonlyArray<string>;
  handlerRegex: RegExp | null;
  propRegex: RegExp | null;
}

const splitPrefixes = (prefixes: string): ReadonlyArray<string> =>
  prefixes
    .split("|")
    .map((token) => token.trim())
    .filter((token) => token.length > 0);

const buildHandlerRegex = (handlerPrefix: string, handlerPropPrefix: string): RegExp | null => {
  if (handlerPrefix.length === 0 || handlerPropPrefix.length === 0) return null;
  const tokens = splitPrefixes(handlerPrefix);
  if (tokens.length === 0) return null;
  const escaped = tokens.map((token) => token.replace(/[.+?^${}()|[\]\\]/g, "\\$&")).join("|");
  return new RegExp(`^((.*\\.)?(${escaped}))[0-9]*[A-Z].*$`);
};

const buildPropRegex = (handlerPropPrefix: string): RegExp | null => {
  if (handlerPropPrefix.length === 0) return null;
  const tokens = splitPrefixes(handlerPropPrefix);
  if (tokens.length === 0) return null;
  const escaped = tokens.map((token) => token.replace(/[.+?^${}()|[\]\\]/g, "\\$&")).join("|");
  return new RegExp(`^(${escaped})[A-Z].*$`);
};

const resolveSettings = (
  settings: Readonly<Record<string, unknown>> | undefined,
): ResolvedSettings => {
  const reactDoctor = settings?.["react-doctor"];
  const ruleSettings =
    typeof reactDoctor === "object" && reactDoctor !== null
      ? ((reactDoctor as { jsxHandlerNames?: JsxHandlerNamesSettings }).jsxHandlerNames ?? {})
      : {};
  let handlerPrefix = DEFAULT_HANDLER_PREFIX;
  let propPrefix = DEFAULT_HANDLER_PROP_PREFIX;
  if (ruleSettings.eventHandlerPrefix === false) handlerPrefix = "";
  else if (typeof ruleSettings.eventHandlerPrefix === "string") {
    handlerPrefix = ruleSettings.eventHandlerPrefix;
  }
  if (ruleSettings.eventHandlerPropPrefix === false) propPrefix = "";
  else if (typeof ruleSettings.eventHandlerPropPrefix === "string") {
    propPrefix = ruleSettings.eventHandlerPropPrefix;
  }
  return {
    checkInlineFunction: ruleSettings.checkInlineFunction ?? false,
    checkLocalVariables: ruleSettings.checkLocalVariables ?? false,
    eventHandlerPrefix: handlerPrefix,
    eventHandlerPropPrefix: propPrefix,
    ignoreComponentNames: ruleSettings.ignoreComponentNames ?? [],
    handlerRegex: buildHandlerRegex(handlerPrefix, propPrefix),
    propRegex: buildPropRegex(propPrefix),
  };
};

// Glob-match supporting `*` only.
// Mirrors OXC's `get_event_handler_name_from_static_member_expression`.
// is_props_handler is TRUE iff the chain is exactly:
//   - `props.<name>` (Identifier "props"), OR
//   - `this.props.<name>` (this.props with `.<name>` once)
// All deeper chains are FALSE.
const getHandlerNameFromMemberExpression = (
  memberExpression: EsTreeNodeOfType<"MemberExpression">,
): { name: string; isPropsHandler: boolean } | null => {
  if (!isNodeOfType(memberExpression.property, "Identifier")) return null;
  const lastName = memberExpression.property.name;
  const object = memberExpression.object as EsTreeNode;
  if (isNodeOfType(object, "Identifier")) {
    return { name: lastName, isPropsHandler: object.name === "props" };
  }
  if (isNodeOfType(object, "MemberExpression")) {
    if (
      isNodeOfType(object.object, "ThisExpression") &&
      isNodeOfType(object.property, "Identifier")
    ) {
      return { name: lastName, isPropsHandler: object.property.name === "props" };
    }
    return { name: lastName, isPropsHandler: false };
  }
  if (isNodeOfType(object, "ThisExpression")) {
    return { name: lastName, isPropsHandler: false };
  }
  return { name: lastName, isPropsHandler: false };
};

// Mirrors OXC's `get_event_handler_name_from_arrow_function`. Arrow
// must have an EXPRESSION body (not block), and its body must be a
// single CallExpression. Returns the call's handler name (callee).
const getHandlerNameFromArrowFunction = (
  arrowFunction: EsTreeNodeOfType<"ArrowFunctionExpression">,
): { name: string; isPropsHandler: boolean } | null => {
  const body = arrowFunction.body;
  // OXC rejects block-body arrows; ESTree marks `expression` field.
  if (isNodeOfType(body, "BlockStatement")) return null;
  if (!isNodeOfType(body, "CallExpression")) return null;
  const callee = body.callee;
  if (isNodeOfType(callee, "Identifier")) {
    return { name: callee.name, isPropsHandler: false };
  }
  if (isNodeOfType(callee, "MemberExpression")) {
    return getHandlerNameFromMemberExpression(callee);
  }
  return null;
};

// "Inline function" check (OXC's `is_member_expression_callee`): true
// iff the arrow's first statement is a CallExpression with a
// member-expression callee. Used to decide whether to walk deeper
// without `checkLocalVariables`. Handles both expression-body and
// block-body arrows.
const isMemberExpressionCallee = (
  arrowFunction: EsTreeNodeOfType<"ArrowFunctionExpression">,
): boolean => {
  const body = arrowFunction.body;
  let callExpression: EsTreeNodeOfType<"CallExpression"> | null = null;
  if (isNodeOfType(body, "CallExpression")) {
    callExpression = body;
  } else if (isNodeOfType(body, "BlockStatement")) {
    const first = body.body[0];
    if (
      first &&
      isNodeOfType(first as EsTreeNode, "ExpressionStatement") &&
      isNodeOfType(
        (first as EsTreeNodeOfType<"ExpressionStatement">).expression as EsTreeNode,
        "CallExpression",
      )
    ) {
      callExpression = (first as EsTreeNodeOfType<"ExpressionStatement">)
        .expression as EsTreeNodeOfType<"CallExpression">;
    }
  }
  return callExpression !== null && isNodeOfType(callExpression.callee, "MemberExpression");
};

// Port of `oxc_linter::rules::react::jsx_handler_names`.
export const jsxHandlerNames = defineRule<Rule>({
  id: "jsx-handler-names",
  title: "Inconsistent event handler names",
  severity: "warn",
  // Stylistic naming convention rule — the upstream pattern
  // (`onClick={handleClick}`) is widely-followed but not universal.
  // The rule also fires on solid-js `<Show when={props.onFoo}>` and
  // similar conditional-render APIs where the `on`-named prop value
  // isn't an event handler. Default off.
  defaultEnabled: false,
  recommendation:
    "Use the `on…` prefix for event-handler props and `handle…` for handlers so readers can trace event flow.",
  category: "Architecture",
  create: (context) => {
    const settings = resolveSettings(context.settings);
    return {
      JSXAttribute(node: EsTreeNodeOfType<"JSXAttribute">) {
        // ignoreComponentNames check.
        if (settings.ignoreComponentNames.length > 0) {
          const opening = node.parent;
          if (!opening || !isNodeOfType(opening as EsTreeNode, "JSXOpeningElement")) return;
          const openingName = (opening as EsTreeNodeOfType<"JSXOpeningElement">).name;
          if (!isNodeOfType(openingName as EsTreeNode, "JSXIdentifier")) return;
          const componentName = (openingName as EsTreeNodeOfType<"JSXIdentifier">).name;
          for (const pattern of settings.ignoreComponentNames) {
            if (compileGlob(pattern).test(componentName)) return;
          }
        }

        const propName = getJsxAttributeName(node.name as EsTreeNode);
        if (!propName) return;
        if (propName === "ref") return;

        const value = node.value as EsTreeNode | null;
        if (!value || !isNodeOfType(value, "JSXExpressionContainer")) return;
        const expression = value.expression as EsTreeNode;
        if (expression.type === "JSXEmptyExpression") return;

        // Determine handler name and whether it's a props-style chain.
        let handlerName: string | null = null;
        let isPropsHandler = false;
        if (isNodeOfType(expression, "MemberExpression")) {
          const result = getHandlerNameFromMemberExpression(expression);
          handlerName = result?.name ?? null;
          isPropsHandler = result?.isPropsHandler ?? false;
        } else if (isNodeOfType(expression, "Identifier")) {
          if (!settings.checkLocalVariables) return;
          handlerName = expression.name;
        } else if (isNodeOfType(expression, "ArrowFunctionExpression")) {
          if (!settings.checkInlineFunction) return;
          if (!settings.checkLocalVariables && !isMemberExpressionCallee(expression)) {
            return;
          }
          const result = getHandlerNameFromArrowFunction(expression);
          handlerName = result?.name ?? null;
          isPropsHandler = result?.isPropsHandler ?? false;
        } else {
          // Other expressions: fall back to literal source if we have it.
          if (!settings.checkLocalVariables) return;
          handlerName = "";
        }

        const propIsEventHandler = settings.propRegex ? settings.propRegex.test(propName) : null;
        let handlerNameIsCorrect: boolean | null;
        if (handlerName === null) {
          handlerNameIsCorrect = false;
        } else if (
          isPropsHandler &&
          settings.propRegex !== null &&
          settings.propRegex.test(handlerName)
        ) {
          handlerNameIsCorrect = true;
        } else if (settings.handlerRegex) {
          handlerNameIsCorrect = settings.handlerRegex.test(handlerName);
        } else {
          handlerNameIsCorrect = null;
        }

        if (propIsEventHandler === true && handlerNameIsCorrect === false) {
          context.report({
            node,
            message: buildHandlerNameMessage(handlerName ?? "", propName),
          });
          return;
        }
        if (propIsEventHandler === false && handlerNameIsCorrect === true) {
          context.report({
            node,
            message: buildHandlerPropMessage(propName, handlerName ?? ""),
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/jsx-handler-names.rs
use oxc_ast::{
    AstKind,
    ast::{
        ArrowFunctionExpression, ComputedMemberExpression, Expression, JSXAttributeValue,
        JSXExpression, Statement, StaticMemberExpression,
    },
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
    utils::get_jsx_attribute_name,
};

use super::simple_glob_matches::simple_glob_matches;

const DEFAULT_HANDLER_PREFIX: &str = "handle";
const DEFAULT_HANDLER_PROP_PREFIX: &str = "on";

#[derive(Debug, Default, Clone)]
pub struct JsxHandlerNames;

struct JsxHandlerNamesSettings {
    check_inline_function: bool,
    check_local_variables: bool,
    handler_prefix: String,
    handler_prop_prefix: String,
    handler_prefixes: Vec<String>,
    handler_prop_prefixes: Vec<String>,
    ignore_component_names: Vec<String>,
}

declare_oxc_lint!(
    /// Enforce consistent event-handler prop and function names.
    JsxHandlerNames,
    react_doctor_native,
    style,
    version = "0.1.0",
    short_description = "Enforce consistent event-handler names.",
);

impl Rule for JsxHandlerNames {
    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        let settings = resolve_jsx_handler_names_settings(ctx);
        for node in ctx.nodes().iter() {
            check_jsx_handler_name(node, &settings, ctx);
        }
    }

    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }
}

fn check_jsx_handler_name<'a>(
    node: &AstNode<'a>,
    settings: &JsxHandlerNamesSettings,
    ctx: &LintContext<'a>,
) {
    let AstKind::JSXAttribute(attribute) = node.kind() else {
        return;
    };
    if !settings.ignore_component_names.is_empty() {
        let AstKind::JSXOpeningElement(opening_element) = ctx.nodes().parent_node(node.id()).kind()
        else {
            return;
        };
        let Some(component_name) = opening_element.name.get_identifier_name() else {
            return;
        };
        if settings
            .ignore_component_names
            .iter()
            .any(|pattern| simple_glob_matches(pattern, component_name.as_str()))
        {
            return;
        }
    }

    let prop_name = get_jsx_attribute_name(&attribute.name);
    if prop_name == "ref" {
        return;
    }
    let Some(JSXAttributeValue::ExpressionContainer(container)) = &attribute.value else {
        return;
    };
    let Some((handler_name, is_props_handler)) =
        resolve_handler_name(&container.expression, settings)
    else {
        return;
    };
    let prop_is_event_handler = handler_prop_name_matches(&prop_name, settings);
    let handler_name_is_correct = handler_name.as_ref().map_or(Some(false), |handler_name| {
        if is_props_handler && handler_prop_name_matches(handler_name, settings) == Some(true) {
            return Some(true);
        }
        handler_name_matches(handler_name, settings)
    });

    match (prop_is_event_handler, handler_name_is_correct) {
        (Some(true), Some(false)) => {
            let handler_name = handler_name.as_deref().unwrap_or_default();
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "The handler \"{handler_name}\" does not match the \"{prop_name}\" event prop convention, so readers cannot trace event flow quickly."
                ))
                .with_label(attribute.span),
            );
        }
        (Some(false), Some(true)) => {
            let handler_name = handler_name.as_deref().unwrap_or_default();
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "The prop \"{prop_name}\" passes handler \"{handler_name}\" but is not named like an event prop, so callers cannot tell it fires an event."
                ))
                .with_label(attribute.span),
            );
        }
        _ => {}
    }
}

fn resolve_handler_name(
    expression: &JSXExpression<'_>,
    settings: &JsxHandlerNamesSettings,
) -> Option<(Option<String>, bool)> {
    match expression {
        JSXExpression::EmptyExpression(_) => None,
        JSXExpression::StaticMemberExpression(member_expression) => {
            let (handler_name, is_props_handler) =
                handler_name_from_static_member_expression(member_expression);
            Some((Some(handler_name.to_string()), is_props_handler))
        }
        JSXExpression::ComputedMemberExpression(member_expression) => Some(
            match handler_name_from_computed_member_expression(member_expression) {
                Some((handler_name, is_props_handler)) => {
                    (Some(handler_name.to_string()), is_props_handler)
                }
                None => (None, false),
            },
        ),
        expression if expression.is_member_expression() => Some((None, false)),
        JSXExpression::Identifier(identifier) => settings
            .check_local_variables
            .then(|| (Some(identifier.name.to_string()), false)),
        JSXExpression::ArrowFunctionExpression(arrow_function) => {
            if !settings.check_inline_function
                || (!settings.check_local_variables && !is_member_expression_callee(arrow_function))
            {
                return None;
            }
            let handler = handler_name_from_arrow_function(arrow_function);
            Some(match handler {
                Some((handler_name, is_props_handler)) => {
                    (Some(handler_name.to_string()), is_props_handler)
                }
                None => (None, false),
            })
        }
        _ if settings.check_local_variables => Some((Some(String::new()), false)),
        _ => None,
    }
}

fn handler_name_from_static_member_expression<'a>(
    member_expression: &'a StaticMemberExpression<'a>,
) -> (&'a str, bool) {
    (
        member_expression.property.name.as_str(),
        is_props_handler_object(&member_expression.object),
    )
}

fn handler_name_from_computed_member_expression<'a>(
    member_expression: &'a ComputedMemberExpression<'a>,
) -> Option<(&'a str, bool)> {
    let Expression::Identifier(identifier) = &member_expression.expression else {
        return None;
    };
    Some((
        identifier.name.as_str(),
        is_props_handler_object(&member_expression.object),
    ))
}

fn is_props_handler_object(object: &Expression<'_>) -> bool {
    match object {
        Expression::Identifier(identifier) => identifier.name == "props",
        Expression::StaticMemberExpression(object_member_expression) => {
            matches!(
                object_member_expression.object,
                Expression::ThisExpression(_)
            ) && object_member_expression.property.name == "props"
        }
        _ => false,
    }
}

fn handler_name_from_arrow_function<'a>(
    arrow_function: &'a ArrowFunctionExpression<'a>,
) -> Option<(&'a str, bool)> {
    let Expression::CallExpression(call_expression) = arrow_function.get_expression()? else {
        return None;
    };
    match &call_expression.callee {
        Expression::Identifier(identifier) => Some((identifier.name.as_str(), false)),
        Expression::StaticMemberExpression(member_expression) => Some(
            handler_name_from_static_member_expression(member_expression),
        ),
        Expression::ComputedMemberExpression(member_expression) => {
            handler_name_from_computed_member_expression(member_expression)
        }
        _ => None,
    }
}

fn is_member_expression_callee(arrow_function: &ArrowFunctionExpression<'_>) -> bool {
    let call_expression = if let Some(Expression::CallExpression(call_expression)) =
        arrow_function.get_expression()
    {
        Some(call_expression.as_ref())
    } else {
        arrow_function
            .get_function_body()
            .and_then(|body| body.statements.first())
            .and_then(|statement| match statement {
                Statement::ExpressionStatement(statement) => match &statement.expression {
                    Expression::CallExpression(call_expression) => Some(call_expression.as_ref()),
                    _ => None,
                },
                _ => None,
            })
    };
    call_expression.is_some_and(|call_expression| call_expression.callee.is_member_expression())
}

fn handler_name_matches(name: &str, settings: &JsxHandlerNamesSettings) -> Option<bool> {
    if settings.handler_prefix.is_empty()
        || settings.handler_prop_prefix.is_empty()
        || settings.handler_prefixes.is_empty()
    {
        return None;
    }
    Some(settings.handler_prefixes.iter().any(|prefix| {
        handler_name_candidate_starts(name, prefix)
            || name
                .match_indices('.')
                .any(|(index, _)| handler_name_candidate_starts(&name[index + 1..], prefix))
    }))
}

fn handler_name_candidate_starts(candidate: &str, prefix: &str) -> bool {
    let Some(remainder) = candidate.strip_prefix(prefix) else {
        return false;
    };
    remainder
        .trim_start_matches(|character: char| character.is_ascii_digit())
        .as_bytes()
        .first()
        .is_some_and(u8::is_ascii_uppercase)
}

fn handler_prop_name_matches(name: &str, settings: &JsxHandlerNamesSettings) -> Option<bool> {
    if settings.handler_prop_prefix.is_empty() || settings.handler_prop_prefixes.is_empty() {
        return None;
    }
    Some(settings.handler_prop_prefixes.iter().any(|prefix| {
        name.strip_prefix(prefix)
            .and_then(|remainder| remainder.as_bytes().first())
            .is_some_and(u8::is_ascii_uppercase)
    }))
}

fn resolve_jsx_handler_names_settings(ctx: &LintContext<'_>) -> JsxHandlerNamesSettings {
    let settings = ctx
        .settings()
        .json
        .as_ref()
        .and_then(|settings| settings.get("react-doctor"))
        .and_then(|settings| settings.get("jsxHandlerNames"))
        .and_then(serde_json::Value::as_object);
    let handler_prefix = configured_handler_prefix(
        settings.and_then(|settings| settings.get("eventHandlerPrefix")),
        DEFAULT_HANDLER_PREFIX,
    );
    let handler_prop_prefix = configured_handler_prefix(
        settings.and_then(|settings| settings.get("eventHandlerPropPrefix")),
        DEFAULT_HANDLER_PROP_PREFIX,
    );
    JsxHandlerNamesSettings {
        check_inline_function: configured_boolean(settings, "checkInlineFunction"),
        check_local_variables: configured_boolean(settings, "checkLocalVariables"),
        handler_prefixes: split_handler_prefixes(&handler_prefix),
        handler_prop_prefixes: split_handler_prefixes(&handler_prop_prefix),
        handler_prefix,
        handler_prop_prefix,
        ignore_component_names: settings
            .and_then(|settings| settings.get("ignoreComponentNames"))
            .and_then(serde_json::Value::as_array)
            .into_iter()
            .flatten()
            .filter_map(serde_json::Value::as_str)
            .map(str::to_string)
            .collect(),
    }
}

fn configured_boolean(
    settings: Option<&serde_json::Map<String, serde_json::Value>>,
    name: &str,
) -> bool {
    settings
        .and_then(|settings| settings.get(name))
        .and_then(serde_json::Value::as_bool)
        .unwrap_or(false)
}

fn configured_handler_prefix(value: Option<&serde_json::Value>, default_prefix: &str) -> String {
    match value {
        Some(serde_json::Value::Bool(false)) => String::new(),
        Some(serde_json::Value::String(prefix)) => prefix.clone(),
        _ => default_prefix.to_string(),
    }
}

fn split_handler_prefixes(prefixes: &str) -> Vec<String> {
    prefixes
        .split('|')
        .map(str::trim)
        .filter(|prefix| !prefix.is_empty())
        .map(str::to_string)
        .collect()
}
